### PR TITLE
[Validator] Fixed calling getters before resolving groups

### LINKED ISSUE #9939: [Validator][GroupSequences] GroupSequences still execute next group if first fail
Hello,

Using the example below (smallest I could think of).

If you try the given controller, you will have the following results:
- test1 : all OK [as expected]
- test2 : "is not three" [as expected]
- test3 : **_CRASH**_ ""Error: Call to a member function myBar() on a non-object in /myProject/src/Me/MeBundle/Doc/Foo.php line 31"" [**_NOT**_ as expected]
- test4 : "bar should not be null" [as expected]

I would expect test3 to give the same result as test4 (Meaning: If the first group of the GroupSequence is violated, the others are not executed).

This is explicitly specified in the doc here : http://symfony.com/doc/current/book/validation.html#group-sequence ""In this example, it will first validate all constraints in the group User (which is the same as the Default group). **Only if all constraints in that group are valid, the second group, Strict, will be validated.**"" (emphasis mine)

and here: https://github.com/symfony/symfony/issues/2947#issuecomment-3509046

cc @bschussek

``` php
<?php

namespace ME\MEBundle\Doc;

use Symfony\Component\Validator\Constraints as Assert;

/**
 * @Assert\GroupSequence({"FilledIn", "Foo"})
 */
class Foo
{
    protected $bar;

    protected $checkBar;

    public function __construct($bar, $checkBar = false)
    {
        $this->bar = $bar;
        $this->checkBar = $checkBar;
    }

    /**
     * @Assert\True(groups={"Foo"}, message="is not three")
     */
    public function isThree()
    {
        if ($this->checkBar && null === $this->bar) {
            return new Bar(null);
        }

        return 3 == $this->getBar()->myBar();
    }

    /**
     * @Assert\NotNull(groups={"FilledIn"}, message="bar should not be null")
     */
    public function getBar()
    {
        return $this->bar;
    }
}
```

``` php
<?php

namespace Me\MeBundle\Doc;

class Bar
{
    protected $bar;

    public function __construct($bar)
    {
        $this->bar = $bar;
    }

    public function myBar()
    {
        return $this->bar;
    }
}
```

``` php
<?php

namespace Me\MeBundle\Controller;

use Me\MeBundle\Doc;

use Symfony\Bundle\FrameworkBundle\Controller\Controller;

use Sensio\Bundle\FrameworkExtraBundle\Configuration\Route;
use Sensio\Bundle\FrameworkExtraBundle\Configuration\Template;

use Symfony\Component\HttpFoundation\JsonResponse;

class DefaultController extends Controller
{
    protected function getFooData(Foo $foo)
    {
        $json = [['bar' => null === $foo->getBar() ? null : (array)$foo->getBar()]];

        $violations = $this->get('validator')->validate($foo);

        $errors = [];

        foreach($violations as $violation) {
            $errors[] = $violation->getMessage();
        }

        $json[] = $errors;

        return $json;
    }

    /**
     * @Route("/test1", defaults={"_format"="json"})
     */
    public function test1Action()
    {
        new Foo(null);

        $foo = new Foo(new Bar(3));

        return new JsonResponse($this->getFooData($foo));
    }

    /**
     * @Route("/test2", defaults={"_format"="json"})
     */
    public function test2Action()
    {
        $foo = new Foo(new Bar(4));

        return new JsonResponse($this->getFooData($foo));
    }

    /**
     * @Route("/test3", defaults={"_format"="json"})
     */
    public function test3Action()
    {
        $foo = new Foo(null);

        return new JsonResponse($this->getFooData($foo));
    }

    /**
     * @Route("/test4", defaults={"_format"="json"})
     */
    public function test4Action()
    {
        $foo = new Foo(null, true);

        return new JsonResponse($this->getFooData($foo));
    }
}
```

### PR DESCRIPTION
**Pull request by HeahDude**
| Q             | A
| ------------- | ---
| Branch?       | 3.4
| Bug fix?      | yes
| New feature?  | no
| Deprecations? | no
| Tickets       | Fix #9939
| License       | MIT
| Doc PR        | ~

### CODE REVIEW

File: src/Symfony/Component/Validator/Validator/PropertyCandidate.php [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+ */
+
+namespace Symfony\Component\Validator\Validator;
+
+/**
+ * A wrapper for a callable initializing a property from a getter.
+ *
+ * @internal
+ */
+class PropertyCandidate
```
Comment (HeahDude) [inline src/Symfony/Component/Validator/Validator/PropertyCandidate.php]: I have trouble naming that class and choosing the right namespace though.

Comment (xabbuh) [reply src/Symfony/Component/Validator/Validator/PropertyCandidate.php]: What about something like `LazyValue`/`LazyPropertyValue`? That would be inline with other places in the framework where we also delay resolving a value.

File: src/Symfony/Component/Validator/Tests/Validator/RecursiveValidatorTest.php [hunk 2]
```diff
@@ -117,6 +121,24 @@ public function testCollectionConstraintValidateAllGroupsForNestedConstraints()
         $this->assertInstanceOf(NotBlank::class, $violations->get(1)->getConstraint());
     }
 
+    public function testGroupedMethodConstraintValidateInSequence()
+    {
+        $metadata = new ClassMetadata('Symfony\Component\Validator\Tests\Fixtures\EntityWithGroupedConstraintOnMethods');
```
Comment (xabbuh) [inline src/Symfony/Component/Validator/Tests/Validator/RecursiveValidatorTest.php]: `EntityWithGroupedConstraintOnMethods::class`?

File: src/Symfony/Component/Validator/Tests/Validator/RecursiveValidatorTest.php [hunk 3]
```diff
@@ -117,6 +121,24 @@ public function testCollectionConstraintValidateAllGroupsForNestedConstraints()
...
+        $this->metadataFactory->addMetadata($metadata);
+
+        $entity = new EntityWithGroupedConstraintOnMethods();
+
+        $metadata->setGroupSequence(['EntityWithGroupedConstraintOnMethods', 'Foo', 'Bar']);
+
+        $violations = $this->validator->validate($entity);
+
+        $this->assertCount(1, $violations);
+        $this->assertInstanceOf('Symfony\Component\Validator\Constraints\IsTrue', $violations->get(0)->getConstraint());
```
Comment (xabbuh) [inline src/Symfony/Component/Validator/Tests/Validator/RecursiveValidatorTest.php]: `IsTrue::class`?

File: src/Symfony/Component/Validator/Validator/PropertyCandidate.php (hunk 1 above)
Comment (HeahDude) [reply src/Symfony/Component/Validator/Validator/PropertyCandidate.php]: Renamed 👍

File: src/Symfony/Component/Validator/Tests/Validator/RecursiveValidatorTest.php (hunk 2 above)
Comment (HeahDude) [reply src/Symfony/Component/Validator/Tests/Validator/RecursiveValidatorTest.php]: Done.

**Review by xabbuh — APPROVED** [review]

**Review by chalasr — APPROVED** [review]

**Comment by nicolas-grekas** [comment]
Thank you @HeahDude.

**Comment by bendavies** [comment]
this is buggy. it breaks traversal of getters returning collections annotated with `Valid` constraint.
I will PR if someone doesn't first.

**Pull request merged**